compiler/rustc_session/src/filesearch.rs

```diff
@@ -62,9 +62,68 @@ pub fn make_target_lib_path(sysroot: &Path, target_triple: &str) -> PathBuf {
     PathBuf::from_iter([sysroot, Path::new(&rustlib_path), Path::new("lib")])
 }
 
-/// This function checks if sysroot is found using env::args().next(), and if it
-/// is not found, uses env::current_exe() to imply sysroot.
-pub fn get_or_default_sysroot() -> PathBuf {
+pub struct SysrootCompileTimeArgsForHost {
+    pub sysroot: Option<String>,
+    pub rustup_home: Option<String>,
+    pub multirust_home: Option<String>,
+    pub rustup_toolchain: Option<String>,
+    pub multirust_toolchain: Option<String>,
+}
+
+/// This function checks if sysroot is found with one of the following way:
+/// 1(if host_args is `None`):
+///     using env::args().next(), and if it is not found, uses env::current_exe() to imply sysroot.
+/// 2(if host_args is `Some`):
+///     Looking from most specific to this invocation to the least:
+///     - runtime environment
+///         - SYSROOT
+///         - RUSTUP_HOME, MULTIRUST_HOME, RUSTUP_TOOLCHAIN, MULTIRUST_TOOLCHAIN
+///     - sysroot from rustc in the path
+///     - compile-time environment
+///         - SYSROOT
+///         - RUSTUP_HOME, MULTIRUST_HOME, RUSTUP_TOOLCHAIN, MULTIRUST_TOOLCHAIN
+pub fn get_or_default_sysroot(host_args: Option<SysrootCompileTimeArgsForHost>) -> Option<PathBuf> {
+    fn toolchain_path(home: Option<String>, toolchain: Option<String>) -> Option<PathBuf> {
+        home.and_then(|home| {
+            toolchain.map(|toolchain| {
+                let mut path = PathBuf::from(home);
+                path.push("toolchains");
+                path.push(toolchain);
+                path
+            })
+        })
+    }
+
+    fn sysroot_from_host(host_args: SysrootCompileTimeArgsForHost) -> Option<PathBuf> {
+        std::env::var("SYSROOT")
+            .ok()
+            .map(PathBuf::from)
+            .or_else(|| {
+                let home =
+                    std::env::var("RUSTUP_HOME").or_else(|_| std::env::var("MULTIRUST_HOME")).ok();
```

Comment:
What if rustup isn't present or RUSTUP_TOOLCHAIN points to a different toolchain? I think moving the sysroot candidate logic of rustc_interface::util here would be better.

Reply:
Why even move it? Couldn't it just be made public in-place?

Reply:
Yes, I mean that one. It would have to be moved as rustc_interface depends on basically every crate, so if it stays in rustc_interface nobody else can use it without cyclic dependencies.

Reply:
sysroot candidate doesn't return the right toolchain either. It contains paths related to build dir. @bjorn3

Reply:
What do you mean? What paths does it return and which are missing?

Reply:
It should likely return all of them and uses should attempt searching for whatever they need in every sysroot candidate.